include/alpaka/atomic/AtomicUniformCudaHip.hpp

```diff
@@ -0,0 +1,474 @@
+/* Copyright 2022 René Widera
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED) || defined(ALPAKA_ACC_GPU_HIP_ENABLED)
+
+#    include <alpaka/atomic/Op.hpp>
+#    include <alpaka/core/BoostPredef.hpp>
+#    include <alpaka/core/Positioning.hpp>
+
+#    include <type_traits>
+
+namespace alpaka
+{
+    //! The GPU CUDA/HIP accelerator atomic ops.
+    //
+    //  Atomics can be used in the hierarchy level grids, blocks and threads.
+    //  Atomics are not guaranteed to be save between devices.
+    class AtomicUniformCudaHipBuiltIn
+    {
+    };
+} // namespace alpaka
+
+#    if !defined(ALPAKA_HOST_ONLY)
+
+#        if defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && !BOOST_LANG_CUDA
+#            error If ALPAKA_ACC_GPU_CUDA_ENABLED is set, the compiler has to support CUDA!
+#        endif
+
+#        if defined(ALPAKA_ACC_GPU_HIP_ENABLED) && !BOOST_LANG_HIP
+#            error If ALPAKA_ACC_GPU_HIP_ENABLED is set, the compiler has to support HIP!
+#        endif
+
+//! Inline namespace is required to check the existenz of function in global namespace via SFINAE.
+inline namespace alpakaGlobal
+{
+    //! Provide an interface to builtin atomic functions.
+    //
+    // To check for the existence of builtin functions located in the global namespace :: directly.
+    // This is forbidden by C++.
+    // If the functor is inheriting from std::false_type an signature is explicitly not available. This can be used to
+    // explicitly disable builtin function in case the builtin is broken.
+    // If the functor is inheriting from std::true_type a spezialization must implement one of the following
+    // interfaces.
+    // \code{.cpp}
+    //    __device__ static T atomic( T* add, T value);
+    //    __device__ static T atomic( T* add, T compare, T value);
+    // \endcode
+    template<typename TOp, typename T, typename THierarchy, typename TSfinae = void, typename TDefer = void>
```

Comment:
I think doxygen comment on template parameters would be helpful here. Perhaps some parts of the comment above could be moved there.